File: backend/app/schemas/search_analytics.py
from typing import Any
import uuid

from pydantic import BaseModel, Field


class AnalyticsMetric(BaseModel):
    value: int | float | None
    label: str
    basis: str


class AnalyticsRate(BaseModel):
    numerator: int
    denominator: int
    rate: float | None
    label: str
    basis: str


class StageConversionMetric(BaseModel):
    from_stage: str
    to_stage: str
    numerator: int
    denominator: int
    rate: float | None
    basis: str


class StageDurationMetric(BaseModel):
    from_stage: str
    to_stage: str
    average_days: float | None
    sample_size: int
    basis: str


class SegmentResponseMetric(BaseModel):
    segment: str
    responses: int
    total: int
    response_rate: float | None
    basis: str


class SearchAnalyticsResponse(BaseModel):
    workspace_id: uuid.UUID | None = None
    scope: str = "all_workspaces"
    summary: dict[str, AnalyticsMetric]
    conversion_rates: list[StageConversionMetric] = Field(default_factory=list)
    average_stage_durations: list[StageDurationMetric] = Field(default_factory=list)
    segment_response_rates: list[SegmentResponseMetric] = Field(default_factory=list)
    signals: list[dict[str, Any]] = Field(default_factory=list)
    insufficient_data: list[str] = Field(default_factory=list)

File: frontend/src/types/searchAnalytics.ts
import type { Insight } from "./insights";

export interface AnalyticsMetric {
  value: number | null;
  label: string;
  basis: string;
}

export interface StageConversionMetric {
  from_stage: string;
  to_stage: string;
  numerator: number;
  denominator: number;
  rate: number | null;
  basis: string;
}

export interface StageDurationMetric {
  from_stage: string;
  to_stage: string;
  average_days: number | null;
  sample_size: number;
  basis: string;
}

export interface SegmentResponseMetric {
  segment: string;
  responses: number;
  total: number;
  response_rate: number | null;
  basis: string;
}

export interface SearchAnalyticsResponse {
  generated_at: string;
  workspace_id: string | null;
  scope: string;
  summary: Record<string, AnalyticsMetric>;
  conversion_rates: StageConversionMetric[];
  average_stage_durations: StageDurationMetric[];
  segment_response_rates: SegmentResponseMetric[];
  signals: Insight[];
  insufficient_data: string[];
}
